File: projet/client_leger/polyquiz/lib/services/game_service.dart
import 'package:audioplayers/audioplayers.dart';
import 'package:flutter/material.dart';
import 'package:polyquiz/constants/socket-event.dart';
import 'package:polyquiz/models/quiz.dart';
import 'package:polyquiz/services/logged_in_user_service.dart';
import 'package:polyquiz/services/offline_game_service.dart';
import 'package:polyquiz/services/real_game_service.dart';
import 'package:polyquiz/services/socket_service.dart';

class GameService extends ChangeNotifier {
  static final GameService _instance = GameService._internal();

  GameService._internal();

  factory GameService() {
    return _instance;
  }

  bool isOfflineMode = false;
  bool isInputFocused = false;
  Map<int, String?> _answers = {};
  String _qrlAnswer = '';
  String get qrlAnswer => _qrlAnswer;
  void set qrlAnswer(String value) {
    this._qrlAnswer = value;
    notifyListeners();
  }
  bool isActive = false;
  bool hasInteracted = false;
  int? lastQrlScore;
  int qreAnswer = 0;
  bool gotNotified = false;
  bool isQuitBtn = false;

  bool isObserverMode = false;
  bool isObservingHost = true;
  String observedUid = '';

  Map<int, String?> get answers {
    if (!this.isObserverMode) return this._answers;
    else return this.realGameService.obsAnswers;
  }

  void set answers(Map<int, String?> value) {
    if (!this.isObserverMode) this._answers = value;
    else this.realGameService.obsAnswers = value;
  }

  int _obsQreAnswer = 0;
  int get obsQreAnswer => this._obsQreAnswer;
  void set obsQreAnswer(int value) {
    this._obsQreAnswer = value;
    realGameService.notifyOnChanged();
    notifyListeners();
  }
  void silentSetObsQre(int value) {
    this._obsQreAnswer = value;
  }
  String get obsQrlAnswer => this.realGameService.obsQrlAnswer;
  void set obsQrlAnswer(String value) {
    this.realGameService.obsQrlAnswer = value;
    notifyListeners();
  }

  final OfflineGameService offlineGameService = OfflineGameService();
  final RealGameService realGameService = RealGameService();
  final SocketService socketService = SocketService();

  int get timer {
    return this.isOfflineMode
        ? this.offlineGameService.timer.time
        : this.realGameService.timer;
  }

  int get playerScore {
    return this.offlineGameService.playerScore;
  }

  bool get isBonus {
    return this.offlineGameService.isBonus;
  }

  QuizQuestion? get question {
    return this.isOfflineMode
        ? this.offlineGameService.question
        : this.realGameService.question;
  }

  QuizQuestion get oldQuestion {
    return this.isOfflineMode
        ? this.offlineGameService.oldQuestion
        : this.realGameService.oldQuestion;
  }

  int get questionNumber {
    return this.isOfflineMode
        ? this.offlineGameService.currQuestionIndex + 1
        : this.realGameService.questionNumber;
  }

  String get username {
    return this.realGameService.username;
  }

  bool get lockedStatus {
    return this.isOfflineMode
        ? this.offlineGameService.locked
        : this.realGameService.locked;
  }

  bool get validatedStatus {
    return this.isOfflineMode
        ? this.offlineGameService.validated
        : this.realGameService.validated;
  }

  AudioPlayer get audio {
    return this.realGameService.audio;
  }

  void destroy() {
    this.reset();
    this.answers.clear();
  }

  void init(String pathId,[bool isObserver=false]) {
    if (isObserver) this.isObserverMode = true;
    // TODO: add configurations for observer
    if (!this.isOfflineMode) {
      configureBaseSockets();
      this.realGameService.roomId = int.parse(pathId);
      this.realGameService.init(this.isObserverMode);
    } else {
      this.offlineGameService.quizId = pathId;
      this.offlineGameService.init();
    }
  }

  void selectChoice(int index) {
    if (!this.lockedStatus) {
      if (this.answers.containsKey(index)) {
        this.answers.remove(index);
        this.realGameService.sendSelection(index, false);
      } else {
        String? textChoice = this.question?.choices?[index].text;
        this.answers[index] = textChoice;
        this.realGameService.sendSelection(index, true);
      }
    }
  }

  void obsUpdateChoice(int index, bool isSelected) {
    if (this.lockedStatus) return;
    if (isSelected && !this.answers.containsKey(index)) {
      String? textChoice = this.question?.choices?[index].text;
      this.answers[index] = textChoice;
      notifyListeners();
      print("After obsUpdateChoice answers are: ${this.answers}");
      return;
    }
    if (!isSelected && this.answers.containsKey(index)) {
      this.answers.remove(index);
      notifyListeners();
    }
    print("After obsUpdateChoice answers are: ${this.answers}\nMind you, we removed a $isSelected answer");
  }

  void selectQREanswer(int selectedAnswer) {
    if (this.lockedStatus || this.isObserverMode) return;
    this.qreAnswer = selectedAnswer;
    this.realGameService.sendQRESelection(selectedAnswer);
  }

  void selectChoiceOffline(int index) {
    if (this.answers.containsKey(index)) {
      this.answers.remove(index);
    } else {
      String? textChoice = this.question?.choices?[index].text;
      this.answers[index] = textChoice;
    }
  }

  void sendAnswer() {
    if (!this.isOfflineMode) {
      this.realGameService.answers = this.answers;
      this.realGameService.qreAnswer = this.qreAnswer;
      this.realGameService.qrlAnswer = this.qrlAnswer;
      this.realGameService.sendAnswer();
      this.isActive = false;
      this.hasInteracted = false;
    } else {
      this.offlineGameService.answers = this.answers;
      this.offlineGameService.qrlAnswer = this.qrlAnswer;
      this.qrlAnswer = '';
      print('ANSWERS: ${this.answers}');
      this.offlineGameService.sendAnswer();
    }
    this.lastQrlScore = null;
    if (!this.isObserverMode) {
      this.socketService.sendMessage(SocketEvent.GET_LAST_QRL_STATUS, {
        'roomId': this.realGameService.roomId,
        'lastQRLScore': this.lastQrlScore,
        'qrlAnswer': this.qreAnswer,
        'userId': LoggedInUserService.instance.user?.uid
      }); // Done
    }
    this.answers.clear();
  }

  bool isPanicDisabled() {
    if (this.question?.type != null) {
      return this.timer < 20 || this.realGameService.inTimeTransition;
    } else {
      return this.timer < 10 || this.realGameService.inTimeTransition;
    }
  }

  void reset() {
    this.isOfflineMode = false;
    this.isInputFocused = false;
    this.qrlAnswer = '';
    this.isActive = false;
    this.hasInteracted = false;
    this.audio.pause();
    this.audio.seek(Duration.zero);
    this.realGameService.destroy();
    this.offlineGameService.reset();
    this.gotNotified = false;
    this.isQuitBtn = false;
    this.isObserverMode = false;
    this.lastQrlScore = null;
    this.qreAnswer = 0;
  }

  void configureBaseSockets() {
    this.socketService.onMessage(SocketEvent.TIME, (timeValue) {
      handleTimeEvent(timeValue);
      notifyListeners();
    });
  }

  void handleTimeEvent(timeValue) {
    this.realGameService.timer = timeValue;
    if (this.timer == 0 && !this.realGameService.locked) {
      // sendQREAnswer not done??
      this.realGameService.locked = true;
      if (this.username != 'host' && !this.isObserverMode) sendAnswer();
    }
  }
}

File: projet/client_lourd/src/app/services/game.service/game.service.ts
import {Injectable} from '@angular/core';
import {QCM_PANIC_MODE_ENABLED, QLR_PANIC_MODE_ENABLED} from '@common/constants/host-interface.component.const';
import {GameRealService} from '@app/services/game-real.service/game-real.service';
import {GameTestService} from '@app/services/game-test.service/game-test.service';
import {SocketClientService} from '@app/services/socket-client.service/socket-client.service';
import {SocketEvent} from '@common/socket-event-name/socket-event-name';
import {HOST_USERNAME} from '@common/names/host-username';
import {QuestionType} from "@common/enums/question-type.enum";
import {Auth} from "@angular/fire/auth";
import {NextQuestionData} from "@common/interfaces/host.interface";
import {TranslateService} from "@ngx-translate/core";
// import {Score} from "@common/interfaces/score.interface";
// import {QuestionStatistics} from "@common/constants/statistic-zone.component.const";

@Injectable({
    providedIn: 'root',
})
export class GameService {
    ///////////////////////////////////
    observerMode: boolean = false;
    observingHost: boolean = true;
    observedPlayerId: string;
    ///////////////////////////////////
    isTestMode: boolean = false;
    isInputFocused: boolean = false;
    answers: Map<number, string | null> = new Map();
    qrlAnswer: string = '';
    isHostEvaluating: boolean = false;
    isActive: boolean = false;
    hasInteracted: boolean = false;
    lastQrlScore: number | undefined = undefined;
    qreAnswer: number;

    // observationValues

    // For player QCM, QRL and QRE answers
    obs_qre_Answer: number;
    obs_qre_interval: string;
    obs_qrl_Answer: string;

    constructor(
        public gameTestService: GameTestService,
        public gameRealService: GameRealService,
        private socketService: SocketClientService,
        private auth: Auth,
        private translate: TranslateService,
    ) {}

    get timer() {
        return this.isTestMode ? this.gameTestService.timer?.time : this.gameRealService.timer;
    }

    get playerScore() {
        return this.gameTestService.playerScore;
    }

    get isBonus() {
        return this.gameTestService.isBonus;
    }

    get question() {
        return this.isTestMode ? this.gameTestService.question : this.gameRealService.question;
    }

    get questionNumber() {
        return this.isTestMode ? this.gameTestService.currQuestionIndex + 1 : this.gameRealService.questionNumber;
    }

    get username() {
        return this.gameRealService.username;
    }

    get lockedStatus() {
        return this.isTestMode ? this.gameTestService.locked : this.gameRealService.locked;
    }

    get validatedStatus() {
        return this.isTestMode ? this.gameTestService.validated : this.gameRealService.validated;
    }

    get audio() {
        return this.gameRealService.audio;
    }

    destroy() {
        this.reset();
        this.answers.clear();
    }

    init(pathId: string, isObserver?: boolean) {
        if (isObserver) this.observerMode = true;
        if (!this.isTestMode) {
            this.configureBaseSockets();
            this.handleGetNextQuestion();
            this.gameRealService.roomId = Number(pathId);
            this.gameRealService.init(isObserver);
        } else {
            this.gameTestService.quizId = pathId;
            this.gameTestService.init();
        }
    }

    selectChoice(index: number) {
        if (!this.lockedStatus) {
            if (this.answers.has(index)) {
                this.answers.delete(index);
                this.gameRealService.sendSelection(index, false);
            } else {
                const textChoice = this.question?.choices ? this.question.choices[index].text : null;
                this.answers.set(index, textChoice);
                this.gameRealService.sendSelection(index, true);
            }
        }
    }

    selectQREAnswer(selectedAnswer: number) {
        if (!this.lockedStatus) {
            this.qreAnswer = selectedAnswer;
            this.gameRealService.qreAnswer = selectedAnswer;
            if (!this.observerMode) this.gameRealService.sendQRESelection(selectedAnswer);
        }
    }

    sendAnswer() {
        if (!this.isTestMode) {
            this.gameRealService.answers = this.answers;
            this.gameRealService.qrlAnswer = this.qrlAnswer;
            this.gameRealService.qreAnswer = this.qreAnswer;
            this.gameRealService.sendAnswer();
            this.isActive = false;
            this.hasInteracted = false;
        } else {
            this.gameTestService.answers = this.answers;
            this.gameTestService.qrlAnswer = this.qrlAnswer;
            this.qrlAnswer = '';
            this.gameTestService.sendAnswer();
        }
        this.lastQrlScore = undefined;
        if (!this.observerMode) {
            this.socketService.send(SocketEvent.GET_LAST_QRL_STATUS, {
                roomId: this.gameRealService.roomId,
                lastQRLScore: this.lastQrlScore,
                qrlAnswer: this.qrlAnswer,
                userId: this.auth.currentUser?.uid,
            })
        }
        this.answers.clear();
    }

    isPanicDisabled() {
        if (this.question?.type == QuestionType.QRL) {
            return this.timer <= QLR_PANIC_MODE_ENABLED || this.gameRealService.inTimeTransition;
        } else {
            return this.timer <= QCM_PANIC_MODE_ENABLED || this.gameRealService.inTimeTransition;
        }
    }

    reset() {
        this.observerMode = false;
        this.isTestMode = false;
        this.qrlAnswer = '';
        this.isActive = false;
        this.hasInteracted = false;
        this.audio.pause();
        this.audio.currentTime = 0;
        this.gameRealService.destroy();
        this.gameTestService.reset();
    }

    private configureBaseSockets() {
        this.socketService.on(SocketEvent.TIME, (timeValue: number) => {
            this.handleTimeEvent(timeValue);
        });
    }

    private handleTimeEvent(timeValue: number) {
        this.gameRealService.timer = timeValue;
        if (this.timer === 0 && !this.gameRealService.locked) {
            if (this.question?.type === QuestionType.QRE && this.username !== HOST_USERNAME) this.selectQREAnswer(this.qreAnswer);
            this.gameRealService.locked = true;
            if (this.username !== HOST_USERNAME && !this.observerMode) this.sendAnswer();
        }
    }

    private handleGetNextQuestion() {
        if (this.socketService.isSocketAlive()) {
            this.socketService.on(SocketEvent.GET_NEXT_QUESTION, (data: NextQuestionData) => {
                if (!this.observerMode) {
                    this.qrlAnswer = "";
                    this.gameRealService.qrlAnswer = "";
                }
                if (this.observerMode) this.qrlAnswer = this.translate.instant('OBSERVER.QRL_PLAYER_INACTIVE');
            });
        }
    }
}
